Custom MyPy binary

When installing dependencies with Poetry, the `mypy ...` command is not available and instead one should use `poetry run mypy ...`.

It would be nice to be able to add an option for the binary used [for the mypy invocation](https://github.com/tsuyoshicho/action-mypy/blob/master/script.sh#L38) (and potentially an option to skip installing mypy without checking if it is installed).

@tsuyoshicho 

Thanks for being so quick to reply!

I just tested the branch, and it works exactly as expected! Actually, I didn't need the `setup_method` change since the `poetry run mypy --version` command succeeds, and installation is skipped even with the `"adaptive"` method.

Also, FYI, I can't run `poetry shell` in GitHub workflows:
```
Run poetry shell
Spawning shell within /home/runner/work/test-stuff/test-stuff/.venv

(25, 'Inappropriate ioctl for device')
Error: Process completed with exit code 1.
```
that's why I have to run `poetry run mypy` explicitly.

--- a/README.md
+++ b/README.md
@@ -45,6 +45,33 @@ inputs:
       Relative to the working directory.
     required: false
     default: '.'
+  ### Flags for setup/execute ###
+  execute_command:
+    description: |
+      mypy execute command.
+      Normally it is "mypy", but for example "poetry run mypy"
+      if you want to run at Poetry without activating the virtual environment.
+    required: false
+    default: 'mypy'
+  setup_command:
+    description: |
+      mypy setup command.
+      Runs when "setup_method" is "install" or required by "adaptive".
+    required: false
+    default: 'pip install mypy'
+  setup_method:
+    description: |
+      mypy setup method. Select from below.
+      "adaptive" - Check "execute_command" with "--version" is executable.
+      If it can be executed, do the same as "nothing", otherwise do the same as "install".
+      "nothing" - no setup process.
+      This option expects the user to prepare the environment
+      (ex. previous workflow step executed "pip install -r requirements.txt").
+      "install" - execute "setup_command".
+
+      Incorrect values behave as "adaptive".
+    required: false
+    default: 'adaptive'
   ### Flags for reviewdog ###
   level:
     description: 'Report level for reviewdog [info,warning,error]'
@@ -116,6 +143,29 @@ jobs:
           # Change the current directory to run mypy command.
           # mypy command reads setup.cfg or other settings file in this path.
           workdir: src
+```
+
+### Using with Poetry
+
+If you use mypy with [Poetry](https://github.com/python-poetry/poetry), you can use it with the following settings (if you don't use `poetry shell`).
+
+```yaml
+name: reviewdog
+on: [pull_request]
+jobs:
+  mypy:
+    name: runner / mypy
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: tsuyoshicho/action-mypy@v3
+        with:
+          github_token: ${{ secrets.github_token }}
+          reporter: github-pr-review
+          level: warning
+          workdir: src
+          execute_command: 'poetry run mypy'
+          setup_method: 'nothing'
 ```
 
 ## Development

--- a/README.md
+++ b/README.md
@@ -147,7 +147,9 @@ jobs:
 
 ### Using with Poetry
 
-If you use mypy with [Poetry](https://github.com/python-poetry/poetry), you can use it with the following settings (if you don't use `poetry shell`).
+If you use mypy with [Poetry](https://github.com/python-poetry/poetry), you can use it with the following settings (`poetry shell` do not work in GitHub workflow, see [issue 66](https://github.com/tsuyoshicho/action-mypy/issues/66)).
+
+Example setting:
 
 ```yaml
 name: reviewdog
@@ -165,7 +167,6 @@ jobs:
           level: warning
           workdir: src
           execute_command: 'poetry run mypy'
-          setup_method: 'nothing'
 ```
 
 ## Development

--- a/action.yml
+++ b/action.yml
@@ -18,6 +18,33 @@ inputs:
       Relative to the working directory.
     required: false
     default: '.'
+  ### Flags for setup/execute ###
+  execute_command:
+    description: |
+      mypy execute command.
+      Normally it is "mypy", but for example "poetry run mypy"
+      if you want to run at Poetry without activating the virtual environment.
+    required: false
+    default: 'mypy'
+  setup_command:
+    description: |
+      mypy setup command.
+      Runs when "setup_method" is "install" or required by "adaptive".
+    required: false
+    default: 'pip install mypy'
+  setup_method:
+    description: |
+      mypy setup method. Select from below.
+      "adaptive" - Check "execute_command" with "--version" is executable.
+      If it can be executed, do the same as "nothing", otherwise do the same as "install".
+      "nothing" - no setup process.
+      This option expects the user to prepare the environment
+      (ex. previous workflow step executed "pip install -r requirements.txt").
+      "install" - execute "setup_command".
+
+      Incorrect values behave as "adaptive".
+    required: false
+    default: 'adaptive'
   ### Flags for reviewdog ###
   level:
     description: 'Report level for reviewdog [info,warning,error]'
@@ -69,6 +96,9 @@ runs:
         INPUT_GITHUB_TOKEN: ${{ inputs.github_token }}
         INPUT_WORKDIR: ${{ inputs.workdir }}
         INPUT_TARGET: ${{ inputs.target }}
+        INPUT_EXECUTE_COMMAND: ${{ inputs.execute_command }}
+        INPUT_SETUP_COMMAND: ${{ inputs.setup_command }}
+        INPUT_SETUP_METHOD: ${{ inputs.setup_method }}
         INPUT_LEVEL: ${{ inputs.level }}
         INPUT_REPORTER: ${{ inputs.reporter }}
         INPUT_FILTER_MODE: ${{ inputs.filter_mode }}

--- a/script.sh
+++ b/script.sh
@@ -11,31 +11,52 @@ echo '::group::🐶 Installing reviewdog ... https://github.com/reviewdog/review
 curl -sfL https://raw.githubusercontent.com/reviewdog/reviewdog/master/install.sh | sh -s -- -b "${TEMP_PATH}" "${REVIEWDOG_VERSION}" 2>&1
 echo '::endgroup::'
 
-echo '::group:: Installing mypy ...  https://github.com/python/mypy'
-if type "mypy" > /dev/null 2>&1 ; then
-  echo 'already installed'
-else
-  echo 'install mypy'
-  pip install mypy
+# check setup method
+SETUP="false"
+case "${INPUT_SETUP_METHOD}" in
+  "nothing")
+    SETUP="false"
+    ;;
+  "install")
+    SETUP="true"
+    ;;
+  *)
+    # adaptive and other invalid value
+    # Check execute_command is valid.
+    echo '::group:: Check command is executable'
+    echo "Execute command with version option: ${INPUT_EXECUTE_COMMAND} --version"
+    if ${INPUT_EXECUTE_COMMAND} --version > /dev/null 2>&1 ; then
+      echo 'Success command execution, skip installation.'
+      SETUP="false"
+    else
+      echo 'Failure command execution, execute installation.'
+      SETUP="true"
+    fi
+    echo '::endgroup::'
+    ;;
+esac
+
+# Install mypy if needed.
+if [[ "${SETUP}" == "true" ]] ; then
+  echo '::group:: Installing mypy ...  https://github.com/python/mypy'
+  echo "Execute setup: ${INPUT_SETUP_COMMAND}"
+  ${INPUT_SETUP_COMMAND}
+  echo '::endgroup::'
 fi
-
-if type "mypy" > /dev/null 2>&1 ; then
-  mypy --version
-else
-  echo 'This repository was not configured for mypy, process done.'
-  exit 1
-fi
-echo '::endgroup::'
-
-export REVIEWDOG_GITHUB_API_TOKEN="${INPUT_GITHUB_TOKEN}"
-
 
 echo '::group:: Running mypy with reviewdog 🐶 ...'
 mypy_exit_val="0"
 reviewdog_exit_val="0"
 
+# Version output.
+echo "Execute command and version: ${INPUT_EXECUTE_COMMAND}"
+${INPUT_EXECUTE_COMMAND} --version
+
+export REVIEWDOG_GITHUB_API_TOKEN="${INPUT_GITHUB_TOKEN}"
+
 # shellcheck disable=SC2086
-mypy_check_output="$(mypy --show-column-numbers     \
+mypy_check_output="$(${INPUT_EXECUTE_COMMAND}       \
+                          --show-column-numbers     \
                           --show-absolute-path      \
                           ${INPUT_MYPY_FLAGS}       \
                           "${INPUT_TARGET:-.}" 2>&1 \

